prog/postgres/postgres_server_nexus.rb

```diff
@@ -279,6 +268,65 @@ def before_run
     COMMAND
 
     vm.sshable.cmd(command)
+    hop_configure_pgbouncer
+  end
+
+  label def configure_pgbouncer
+    pgbouncer_password = SecureRandom.urlsafe_base64(15)
+
+    pgbouncer_setup_query = <<~PGBOUNCER_SETUP
+      /**
+       * Create pgbouncer role if it does not exist.
+       */
+      DO $$
+      BEGIN
+          IF NOT EXISTS (SELECT 1 FROM pg_roles WHERE rolname = 'pgbouncer') THEN
+              CREATE ROLE pgbouncer LOGIN ENCRYPTED PASSWORD '#{pgbouncer_password}';
+          END IF;
+      END
+      $$;
+
+      /**
+       * Lock down the privileges of the pgbouncer role.
+       */
+      REVOKE ALL PRIVILEGES ON SCHEMA public FROM pgbouncer;
+
+      /**
+       * Create the pgbouncer schema if it does not exist. All of the
+       * administrative functions for pgbouncer will live in its own schema.
+       */
+      CREATE SCHEMA IF NOT EXISTS pgbouncer;
+
+      /**
+       * Lock down the privileges of the pgbouncer schema.
+       */
+      REVOKE ALL PRIVILEGES ON SCHEMA pgbouncer FROM pgbouncer;
+      GRANT USAGE ON SCHEMA pgbouncer TO pgbouncer;
+
+      /**
+       * The "get_auth" function is used by pgbouncer to authenticate users.
+       * See: http://www.pgbouncer.org/config.html#auth_query
+       */
+      CREATE OR REPLACE FUNCTION pgbouncer.get_auth (
+        INOUT p_user     name,
+        OUT   p_password text
+      ) RETURNS record
+        LANGUAGE sql SECURITY DEFINER SET search_path = pg_catalog AS
+      $$SELECT usename, passwd FROM pg_shadow WHERE usename = p_user$$;
+
+      REVOKE ALL ON FUNCTION pgbouncer.get_auth(name) FROM PUBLIC, pgbouncer;
+      GRANT EXECUTE ON FUNCTION pgbouncer.get_auth(name) TO pgbouncer;
+    PGBOUNCER_SETUP
+
+    pgbouncer_setup_result = postgres_server.run_query(pgbouncer_setup_query).chomp
```

Comment:
is this transactional, e.g. is `-1` implied? If not, bracketing the whole text in BEGIN/COMMIT would be better if possible.